From 7b8bb36aff6cd4eae9a6a24b8fdd359b8664f0a0 Mon Sep 17 00:00:00 2001
From: Davidson K <davidson.kumaresan@arm.com>
Date: Thu, 23 Jun 2022 11:24:01 +0530
Subject: [PATCH 07/14] fwu: use image type uuid from the gpt information

Associating the image type uuid to the partition list helps in
retrieving the partition details associated with the image type
uuid.

Signed-off-by: Davidson K <davidson.kumaresan@arm.com>
Change-Id: Ic1c2ad0367a4b1a6a5d1bfb71f4b26648e9fdc76
---
 platform/drivers/arm/partition/gpt.c       |  1 +
 platform/drivers/arm/partition/partition.c | 15 +++++++++++++++
 platform/interface/partition/partition.h   |  2 ++
 3 files changed, 18 insertions(+)

diff --git a/platform/drivers/arm/partition/gpt.c b/platform/drivers/arm/partition/gpt.c
index 3a45613..36e07f2 100644
--- a/platform/drivers/arm/partition/gpt.c
+++ b/platform/drivers/arm/partition/gpt.c
@@ -57,6 +57,7 @@ int parse_gpt_entry(gpt_entry_t *gpt_entry, partition_entry_t *entry)
 				   gpt_entry->first_lba + 1) *
 			PLAT_PARTITION_BLOCK_SIZE;
 	memcpy(&entry->part_guid, &gpt_entry->unique_uuid, sizeof(struct efi_guid));
+	memcpy(&entry->type_guid, &gpt_entry->type_uuid, sizeof(struct efi_guid));
 
 	return 0;
 }
diff --git a/platform/drivers/arm/partition/partition.c b/platform/drivers/arm/partition/partition.c
index a5f871d..2a2cfbd 100644
--- a/platform/drivers/arm/partition/partition.c
+++ b/platform/drivers/arm/partition/partition.c
@@ -265,6 +265,21 @@ const partition_entry_t *get_partition_entry_by_uuid(const uuid_t *part_uuid)
 	return NULL;
 }
 
+const partition_entry_t *get_partition_entry_by_type_uuid_and_index(const uuid_t *type_uuid, int index)
+{
+	int i;
+	int found = 0;
+
+	for (i = 0; i < list.entry_count; i++) {
+		if (memcmp(type_uuid, &list.list[i].type_guid, sizeof(uuid_t)) == 0) {
+			if(found++ == index)
+				return &list.list[i];
+		}
+	}
+
+	return NULL;
+}
+
 const partition_entry_list_t *get_partition_entry_list(void)
 {
 	return &list;
diff --git a/platform/interface/partition/partition.h b/platform/interface/partition/partition.h
index 64d0ab9..7987065 100644
--- a/platform/interface/partition/partition.h
+++ b/platform/interface/partition/partition.h
@@ -38,6 +38,7 @@ typedef struct partition_entry {
 	uint64_t		length;
 	char			name[EFI_NAMELEN];
 	struct efi_guid		part_guid;
+	struct efi_guid		type_guid;
 } partition_entry_t;
 
 typedef struct partition_entry_list {
@@ -48,6 +49,7 @@ typedef struct partition_entry_list {
 int load_partition_table(void);
 const partition_entry_t *get_partition_entry(const char *name);
 const partition_entry_t *get_partition_entry_by_uuid(const uuid_t *part_uuid);
+const partition_entry_t *get_partition_entry_by_type_uuid_and_index(const uuid_t *part_uuid, int index);
 const partition_entry_list_t *get_partition_entry_list(void);
 int partition_init(void);
 
-- 
2.34.1

